G-code console for LVGL UI

X-Ryl669 opened the pull request:
### Description

This is an additional menu for LVGL based UI to allow inputing G-Code directly in the interface and getting answer
It's located in the "More" menu from the "Tool" menu. The flow is like this:
1. Select the G-Code option in the More menu
![IMG_20210112_071906_HHT](https://user-images.githubusercontent.com/3277165/104294391-654d8d00-54bf-11eb-9e9e-6b8909c8ae38.jpg)
2. Click on "Enter G-Code" and enter your G-Code command
![IMG_20210112_071927](https://user-images.githubusercontent.com/3277165/104294402-67afe700-54bf-11eb-9724-1ff58b3bf673.jpg)
3. Upon return, the result of the G-Code command is displayed
![IMG_20210112_071936_HHT](https://user-images.githubusercontent.com/3277165/104294408-68e11400-54bf-11eb-8a0a-f50feffc3bbc.jpg)

I don't expect this code to land as-is, but instead request for comments about how to improve it or to better integrate with it.


### Requirements

This requires LVGL, and I've based this code above the pending PR #20191.
I hope the former will be merged soon and I think it's better not to append more UI change on this other PR. 

### Benefits

This opens a lot of possibilities to use your printer without a computer connected to it (in standalone mode). All the features of Marlin are now accessible via the LCD screen (albeit via technical & textual representation) even if a graphical menu is missing to deal with it (for example, getting the current position of all axis is not present by default in LVGL, and this is useful to calibrate a BLTouch sensor).

### Configurations

No configuration required. It's enabled by default when LVGL is enabled.

### Issues
Currently, as you'll discover, I had to add a hook to the serial's `write` function to capture the output of a G-Code command. 
Looking at the code, there's already some kind of hooking function for ethernet but it's completely non-expandable for third party.
I think Marlin would benefit from a better hooking system (something more generic to capture anything that goes to serial output whatever the serial output selected). Looking at the code, this means either rewriting the `SERIAL_CHAR` et al. macros so that they can be switched at compile time and/or runtime. 

I need to discuss with actual developers what would be the best solution. The solution in this PR is working for MarlinSerial's based on Hardware Serial, but not on USB serial, so it's far from perfect. It's using a pointer to a callback function that can be swapped to a null value to disable/enable hooking at runtime.

Another issue I have is to detect the end of a GCode command's answer. Some command ends with "ok" and I'm currently using this as a sentinel to unhook the serial capture, but I wonder if a signal could be added in the G-code processor to tells any hook that a command has completed processing.
 
Please let me know what you think, thanks!

makerbase-mks committed:
Merge pull request #1 from MarlinFirmware/bugfix-2.0.x

Bugfix 2.0.x

makerbase-mks committed:
Merge pull request #2 from MarlinFirmware/bugfix-2.0.x

Bugfix 2.0.x

makerbase-mks committed:
Merge pull request #4 from MarlinFirmware/bugfix-2.0.x

Bugfix 2.0.x

makerbase-mks committed:
Merge pull request #6 from MarlinFirmware/bugfix-2.0.x

Bugfix 2.0.x

makerbase-mks committed:
Merge pull request #7 from MarlinFirmware/bugfix-2.0.x

Bugfix 2.0.x

makerbase-mks committed:
Merge pull request #14 from MarlinFirmware/bugfix-2.0.x

Bugfix 2.0.x

makerbase-mks committed:
Merge pull request #15 from MarlinFirmware/bugfix-2.0.x

Bugfix 2.0.x

MKS-Sean committed:
Merge pull request #16 from MarlinFirmware/bugfix-2.0.x

Bugfix 2.0.x

MKS-Sean committed:
Merge pull request #27 from MarlinFirmware/bugfix-2.0.x

Bugfix 2.0.x

MKS-Sean committed:
Increase the MKS WI-FI module function of TFT_LVGL_UI

thinkyhead committed:
Clean up formatting

thinkyhead committed:
Fix formatting, option name

thinkyhead committed:
also

thinkyhead committed:
Fix typos

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/20191

thinkyhead committed:
Cleanup

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/20191

thinkyhead committed:
Misc cleanup

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/20191

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/20191

thinkyhead committed:
Fix indent

rhapsodyv committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/20191

rhapsodyv committed:
use uint8_t for binary data

rhapsodyv committed:
minor cleanup

rhapsodyv committed:
save some bytes

rhapsodyv committed:
fix typo

rhapsodyv committed:
more cleanup

thinkyhead committed:
mks-ui cleanup

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/20191

thinkyhead committed:
buf => cmd_wifi_list

thinkyhead committed:
match types / sizes

thinkyhead committed:
spacing and spelling

thinkyhead committed:
Consolidate buffers

thinkyhead committed:
Fix sign on babystep_dist

thinkyhead committed:
Exit if cur_name is null

MKS-Sean committed:
Merge the revised code of X-Ryl669

MKS-Sean committed:
Restore default configuration

MKS-Sean committed:
Delete the comment code and modify the code format

X-Ryl669 committed:
Fix bugs in the file selector screen and icon not being displayed correctly

MKS-Sean committed:
Resolve conflicting files

MKS-Sean committed:
Resolve conflicting files

MKS-Sean committed:
lvgl ui adds calibration function

MKS-Sean committed:
Resolve file conflicts

MKS-Sean committed:
Merge pull request #29 from MarlinFirmware/bugfix-2.0.x

Bugfix 2.0.x

makerbase-mks committed:
Merge pull request #30 from MarlinFirmware/bugfix-2.0.x

Update Bugfix 2.0.x

X-Ryl669 committed:
Adding current position to the move menu (not finished yet)
Add a report of the current position to the move page so it's easier to get information from

X-Ryl669 committed:
Fix missing stack return causing infinite usage of display stack

X-Ryl669 committed:
Fix autoleveling clearing screen when not appropriate
Fix title being cleared when it's too long because the first item in stack does not produce a page text

Rockman18 committed:
[FIX] Case sensitive include on Linux

MKS-Sean committed:
Merge pull request #34 from Rockman18/fix-wifi

[FIX] Case sensitive include on Linux

MKS-Sean committed:
Merge pull request #33 from X-Ryl669/AddStatusInMoveMenu

Add status in move menu

MKS-Sean committed:
Merge pull request #31 from X-Ryl669/FixIcon

Fix bugs in the file selector screen and icon not being displayed correctly

MKS-Sean committed:
Fix CI check problem

MKS-Sean committed:
public_bufpublic_buf

MKS-Sean committed:
resolve conflict

MKS-Sean committed:
resolve conflicts

MKS-Sean committed:
resolve conflicts

MKS-Sean committed:
Fix wifi transmission problem

MKS-Sean committed:
resolve conflicts

MKS-Sean committed:
resolve conflicts

rhapsodyv committed:
file added by mistake?

rhapsodyv committed:
Keep comments for future test env

rhapsodyv committed:
obsolete tft code

X-Ryl669 committed:
Fix bug reported by Malderin about motor off

X-Ryl669 committed:
WIP: Add GCode input method to the graphical menu (so the printer can be used without a computer)

X-Ryl669 committed:
WIP: Add GCode dialog to UI so it's possible to enter GCode command directly via touch screen and get the answer

X-Ryl669 commented:
<failure in CI not my code>

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/20755

thinkyhead commented:
This seems to be missing the code for `lv_draw_gcode` and the `draw_gcode.h` file.

thinkyhead committed:
Merge bugfix-2.0.x into pr/20755

thinkyhead commented on Marlin/src/lcd/extui/lib/mks_ui/tft_lvgl_configuration.cpp:
```diff
@@ -120,6 +118,10 @@ void tft_lvgl_init() {
   watchdog_refresh();     // LVGL init takes time
 
+  #if MB(MKS_ROBIN_NANO)
+    OUT_WRITE(PB0, LOW);  // HE1
```
Sometimes this is the `FAN1_PIN` and sometimes `HEATER_1_PIN`. What LVGL-related function does this line serve?

X-Ryl669 replied:
It's not my code. It's in #20191. On my board, it's HEATER2 pin. *Maybe* it's there to disable heating element on boot? I don't think it's required, but we should ask @MKS-Sean for more details.

MKS-Sean replied:
In our actual test, the ROBIN_NANO_V1.2 motherboard will be opened when heater2 is started. It is dangerous to open it when updating assets

thinkyhead replied:
> "the ROBIN_NANO_V1.2 motherboard will be opened…"

What do you mean by "**opened**" here?

MKS-Sean replied:
Sorry, I didn’t express it clearly. PB0 will be set to high level, causing the MOS FET connected to it to turn on

thinkyhead committed:
General cleanup

X-Ryl669 committed:
Add missing files

X-Ryl669 commented:
> This seems to be missing the code for `lv_draw_gcode` and the `draw_gcode.h` file.

Fixed, sorry.

rhapsodyv commented:
This PR should be merged only after #20191, right?

It’s becaming a little mess. Now we have two PR in 99% equal.
And one third that was merged as base of the these two....

Please, set this as draft.

@thinkyhead please, merge #20191. There’s no reason for hold it.

thinkyhead commented:
> please, merge #20191. There’s no reason for hold it.

It has a long discussion thread, and at the end the last comment indicates that someone wants do some testing and report back. So if you want to help move that PR forward, please add commentary there about your own review and get consensus from the discussion participants that it is okay to go forward. I will take care of bringing it up to date and resolving conflicts so that it is in proper shape and has parity with the latest codebase.

rhapsodyv commented:
@thinkyhead 

I did: https://github.com/MarlinFirmware/Marlin/pull/20191#issuecomment-757553256

And fixed the main issue: https://github.com/MarlinFirmware/Marlin/pull/20191#issuecomment-757542854

I tested everything already: https://github.com/MarlinFirmware/Marlin/pull/20191#issuecomment-740205661

And even asked mks to stop adding random things here: https://github.com/MarlinFirmware/Marlin/pull/20191#issuecomment-751554071

And mks tested it again: https://github.com/MarlinFirmware/Marlin/pull/20191#issuecomment-752355507

thinkyhead commented:
Cleanup is the main thing eating up time on PRs like these. I have already put in several hours and there is still more to do. I will put in as much time as needed and then I will go ahead and merge that PR according to your recommendation.

thinkyhead committed:
MKS UI general cleanup

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/20755

thinkyhead committed:
more cleanup

X-Ryl669 commented:
Now that #20191 is merged, do you want me to drop this pr and reopen a new one with the only 3 commits ahead of bugfix ?

X-Ryl669 commented:
> This PR should be merged only after #20191, right?

No, I need some guide about how to do things correctly. Currently, the code is working on my board, but since I've only modified MarlinSerial for STM32F1 it'll not work for USB serial for example. 
I think a more general solution would be required, and I'm "requesting for comment".  

Once we are converging on a technical solution, I'll update the PR and implement it.

thinkyhead committed:
minor tweaks

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/20755

thinkyhead committed:
Misc. g-code injection fixes

thinkyhead committed:
inline G91/G90

thinkyhead committed:
Use G90/G91

thinkyhead commented:
You don't need to worry about squashing your commits. We can just look at the cumulative changes in our review for something like this. It's pretty cleaned up at this point so it's just down to whether it actually functions properly now!

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/20755

thinkyhead committed:
Use memcmp_P

thinkyhead committed:
Add memcmp_P to HALs

thinkyhead committed:
outdent

X-Ryl669 commented:
@thinkyhead What is the point of adding `_P` version of the strcmp/memcmp et al ? 
If I understand correctly, LVGL does not work on AVR, and all other platforms don't have specific requirement for such functions, don't they ? 

I'll try another (generic) approach for hooking the serial output so you can comment on it. Hopefully, it'll work on all platform.

rhapsodyv commented:
I think that we must avoid adding any UI code in the marlin core. That being said, to do what you want, I think the clear and simple way is just make `MarlinSerial` accept a `onWriteCallback` (or so), that clients can attach/un-attach, to recebe what is being send to serial.

```cpp
class MarlinSerial : public FrameworkSerial {
public:
  void setOnWriteCallback(void(*callback)(const char*)) { onWriteCallback = calback; }

private:
  void write(data) {
    if (onWriteCallback) onWriteCallback(data);
    FrameworkSerial::write(data);
  }
 onWriteCallback;
};

/// UI
void myScreenSerialWriteCallback(data) {
   //write the data to the screen
}

void my_screen() {
   MarlinSerial1.setOnWriteCallback(myScreenSerialWriteCallback);
}
```

X-Ryl669 committed:
Start implementing a generic serial hook class.
This is currently not finished.

The idea is to have the compiler optimize all the hook functions when none are being used without relying on a preprocessor configuration hell.
All of this code will be optimized away when built so it should not impact the binary size nor the runtime speed.
Once finished, it should just work as expected while simplifying the current serial.h file.

This will allow to have as many serial port as desired, and, hopefully, should also permit to remove many redundant code in the different HAL, while hooking any serial port (for GUI console or telnet or whatever).

X-Ryl669 commented:
@rhapsodyv The first code worked this way. It's simple but it requires changing all HAL code to have a callback in there, so you pay the cost of testing for a hook function even if you don't have a LVGL UI. You are right that it's quite simple.

In the current work in progress (not working yet), I'm making the hook code selection at compile time, so you don't pay the cost if your configuration does not require it. I've rewritten a bit of serial's wizardry macros in serial.h and serial_hook.h, and hopefully it'll be more simple to follow.  

The idea will be to build a chain of SerialHook instance based on the configuration, it can be as simple as BaseSerial (what we have now), or RuntimeSerial (with a runtime modifiable hook like you've written) or MultiSerial which output to 2 output (both are configurable). All of this with zero overhead (no virtual table, no test for hook function pointer if none required), so it compiles to the bare minimum.

rhapsodyv commented:
Just enable when it is used. No other Hal need be changed:

```cpp
class MarlinSerial : public FrameworkSerial {
public:

#if HAS_SERIAL_WRITE_HOOK
  void setOnWriteCallback(void(*callback)(const char*)) { onWriteCallback = calback; }

private:
  void write(data) {
    if (onWriteCallback) onWriteCallback(data);
    FrameworkSerial::write(data);
  }
 onWriteCallback;
#endif
};

/// UI
void myScreenSerialWriteCallback(data) {
   //write the data to the screen
}

void my_screen() {
   MarlinSerial1.setOnWriteCallback(myScreenSerialWriteCallback);
}


Stm32f1 condicionals_post.h:
#if ENABLED(TFT_LVGL_UI)
   #define HAS_SERIAL_WRITE_HOOK
#endif 


```

This may solve the issue and avoid the need for any complex classes.

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/20755

thinkyhead committed:
General serial cleanup

thinkyhead committed:
etc.

rhapsodyv commented:
@thinkyhead Better take a look on #20783

X-Ryl669 commented:
Indeed, sorry Scott, I've followed Victor's advice and split the PR in 2 part, the serial hook stuff in #20783 and this code that is postponed until the serial change is done (and I'll update the code to use the new serial hook stuff).

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/20755

thinkyhead committed:
merge followup

thinkyhead committed:
comment

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/20755

X-Ryl669 commented:
I'll resume working on this since the serial code change is merged. I'll let you know when it's done. Thanks!

thinkyhead committed:
[cron] Bump distribution date (2021-01-29)

ellensp committed:
Fix LED_CONTROL_MENU compile (#20914)

ellensp committed:
Fix undefined G28_STR (#20912)

GerogeFu committed:
FYSETC Cheetah 2.0 (#20897)

X-Ryl669 committed:
Optimize serial output code for size (#20911)

thinkyhead committed:
AutoReport class (Temperature, Cardreader) (#20913)

rafaljot committed:
MPX_ARM_MINI board (Mingda MD-16) (#20711)

thinkyhead committed:
[cron] Bump distribution date (2021-01-30)

ellensp committed:
Fix G28_STR (#20925)

ellensp committed:
KILL, BEEPER pins for LCD_FOR_MELZI (#20924)

ConstantijnCrijnen committed:
Language selection auto-save option (#20915)

X-Ryl669 committed:
Fix merge issue

X-Ryl669 committed:
Update to use new serial hook code

X-Ryl669 committed:
Merge branch 'bugfix-2.0.x' into AddGCodeMenu

X-Ryl669 commented:
Should be ready now for inclusion...

thinkyhead committed:
Clean up

Valnar8 commented:
I like the idea but I always get a blackscreen after pressing the check mark. I guess that's what you meant with the end detection issue. Even ok didn't help in my case.

However I have some more feedback to this: It's a bit of a pain to enter g-code with this Keyboard layout. Would be a lot better if the standard screen was a num-block, a dot and the most important letters (G, M, X, Y, Z, E and what else are common letters for G-code Parameters)
